### PR TITLE
[vcpkg-tool][world rebuild] Instrument vcpkg.cmake to check for optional dependencies

### LINKED ISSUE #27590: [vcpkg-tool] Post build lint: Check for missing optional dependencies
**Is your feature request related to a problem? Please describe.**

Many ports have optional dependencies on packages that are either provided by vcpkg or a system package manager. This is the case if upstream uses `find_package()` without `REQUIRED` and reacts differently if the package is found. Dependencies are either unexpectedly 'picked up' or not found [this is especially relevant for ports that use the `unofficial::` namespace]. This leads to unreproducible builds or - even worse - to build failures.

In orther to mitigate this, we have to explicitly control dependencies via `CMAKE_DISABLE_FIND_PACKAGE_` and `CMAKE_REQUIRE_FIND_PACKAGE_`. Unfortunately, we almost never notice such behavior because the package is either not available in our CI or always installed before the port that depends on it.

**Proposed solution**

Add a post build check that searches for `Could NOT find <package name> (missing ...` in `config-<triplet>-out.log`.

### PR DESCRIPTION
**Pull request by Thomas1664**
**Describe the pull request**

- #### What does your PR fix?
  Fixes #27590

See the tagged issue for more information.

This PR adds a check to the vcpkg override of `find_package` to detect if it is called without `REQUIRED` and not explicitly handled using `CMAKE_DISABLE_FIND_PACKAGE_` or `CMAKE_REQUIRE_FIND_PACKAGE_`.
It needs a PR on the tool as well.

### CODE REVIEW

**Comment by Thomas1664** [comment]
> Did you test this locally, with non-trivial ports/projects? I think this is too simple with regard to recursive `find_package/find_dependency` calls. In particular the latter must never use `REQUIRED` but has to check `<Pkg>_FOUND`.

With the limited resources of a Codespace I tested ports like sdl2 or OSG successfully. Can you recommend some ports that fin in 20 GB of storage and build in a reasonable amount of time?

**Comment by Thomas1664** [comment]
> `qtbase`. Probably already breaks with minimal features.... But from experience I can tell you that this is a waste of time. I tried to force `CMAKE_REQUIRE_FIND_PACKAGE_` in the qt ports but qt seems to call `find_package` first with the `CONFIG` keyword and after that without that which is perfectly fine for CMake, e.g.
> 
> ```
> find_package(package CONFIG)
> find_package(package REQUIRED)
> ```
> 
> and you cannot deal with that correctly. The same goes for `COMPONENTS` and `OPTIONAL_COMPONENTS`

Since `find_package` is a macro we might be able to record the package name between calls?

Even if some ports break I think the majority of ports is simple enough to work with this. We can set a flag to disable this check for some complex ports like qt.

**Comment by Neumann-A** [comment]
> Since find_package is a macro we might be able to record the package name between calls?

What about recursive and nested find_package calls ?

> Even if some ports break I think the majority of ports is simple enough to work with this. 

user code also interacts with `vcpkg.cmake` so it needs to be off by default. Also you could just inject cmake_language(DEFER) and check the property https://cmake.org/cmake/help/latest/prop_gbl/PACKAGES_NOT_FOUND.html

**Comment by Thomas1664** [comment]
My new approach based on @Neumann-A's suggestions:

- Define a variable outside vcpkg's `find_package` that stores the package names ans relevant arguments like `CONFIG`, `REQUIRED`, `(OPTIONAL_)COMPONENTS`
- Add a package on each invocation of `find_package` to this list [nested/recursive calls shouldn't matter?]
- Use `cmake_language(DEFER)` to process `<PackageName>_FIND_REQUIRED` and `<PackageName>_FIND_REQUIRED_<c>`: https://cmake.org/cmake/help/latest/command/find_package.html#package-file-interface-variables
- Issue a warning if a package is inside `PACKAGES_NOT_FOUND` or <PackageName>_FIND_REQUIRED` / `<PackageName>_FIND_REQUIRED_<c>` are false

Only enable this check for vcpkg internal usage.

File: scripts/buildsystems/vcpkg.cmake [hunk 1]
```diff
@@ -757,6 +757,12 @@ macro("${VCPKG_OVERRIDE_FIND_PACKAGE_NAME}" z_vcpkg_find_package_package_name)
     set(z_vcpkg_find_package_ARGN "${ARGN}")
     set(z_vcpkg_find_package_backup_vars)
 
+    string(REGEX MATCH "^REQUIRED$" z_vcpkg_package_required "${z_vcpkg_find_package_ARGN}")
```
Comment (Thomas1664) [inline scripts/buildsystems/vcpkg.cmake:760]: The regex is also wrong

File: scripts/buildsystems/vcpkg.cmake [hunk 2]
```diff
@@ -757,6 +757,12 @@ macro("${VCPKG_OVERRIDE_FIND_PACKAGE_NAME}" z_vcpkg_find_package_package_name)
...
+
+    if(NOT z_vcpkg_package_required AND NOT (CMAKE_DISABLE_FIND_PACKAGE_${z_vcpkg_find_package_package_name} OR CMAKE_REQUIRE_FIND_PACKAGE_${z_vcpkg_find_package_package_name}))
+        message(Warning "[VCPKG_IMPLICIT_DEPENDENCY] ${z_vcpkg_find_package_package_name} is not controlled explicitly.")
```
Comment (Thomas1664) [inline scripts/buildsystems/vcpkg.cmake:763]: must be `WARNING`

**Comment by Thomas1664** [comment]
Depends on #27950 and #27982

**Comment by BillyONeal** [comment]
I believe this PR is replaced by https://github.com/microsoft/vcpkg/pull/27950 . Thanks for your contribution to vcpkg!

**Pull request closed without merging**